File: src/behaviortree/nodes/conditions/condition.cpp
/////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// Tencent is pleased to support the open source community by making behaviac available.
//
// Copyright (C) 2015-2017 THL A29 Limited, a Tencent company. All rights reserved.
//
// Licensed under the BSD 3-Clause License (the "License"); you may not use this file except in compliance with
// the License. You may obtain a copy of the License at http://opensource.org/licenses/BSD-3-Clause
//
// Unless required by applicable law or agreed to in writing, software distributed under the License is
// distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and limitations under the License.
/////////////////////////////////////////////////////////////////////////////////////////////////////////////////

#include "behaviac/common/base.h"
#include "behaviac/behaviortree/nodes/conditions/condition.h"

#include "behaviac/common/profiler/profiler.h"
#include "behaviac/behaviortree/nodes/actions/action.h"
#include "behaviac/common/meta.h"

namespace behaviac {
    Condition::Condition() : m_opl(0), m_opr(0), m_operator(E_EQUAL) {
    }

    Condition::~Condition() {
        BEHAVIAC_DELETE(m_opl);
        BEHAVIAC_DELETE(m_opr);
    }

    void Condition::load(int version, const char* agentType, const properties_t& properties) {
        super::load(version, agentType, properties);

        behaviac::string typeName;

        for (propertie_const_iterator_t it = properties.begin(); it != properties.end(); ++it) {
            const property_t& p = (*it);

            if (StringUtils::StringEqual(p.name, "Operator")) {
                this->m_operator = OperationUtils::ParseOperatorType(p.value);
            } else if (StringUtils::StringEqual(p.name, "Opl")) {
                const char* pParenthesis = strchr(p.value, '(');

                if (pParenthesis == 0) {
                    this->m_opl = AgentMeta::ParseProperty(p.value);
                } else {
                    this->m_opl = AgentMeta::ParseMethod(p.value);
                }
            } else if (StringUtils::StringEqual(p.name, "Opr")) {
                const char* pParenthesis = strchr(p.value, '(');

                if (pParenthesis == 0) {
                    this->m_opr = AgentMeta::ParseProperty(p.value);
                } else {
                    this->m_opr = AgentMeta::ParseMethod(p.value);
                }
            } else {
                BEHAVIAC_ASSERT(0, "unrecognised property %s", p.name);
            }
        }

    }

    bool Condition::IsValid(Agent* pAgent, BehaviorTask* pTask) const {
        if (!Condition::DynamicCast(pTask->GetNode())) {
            return false;
        }

        return super::IsValid(pAgent, pTask);
    }

    bool Condition::Evaluate(Agent* pAgent) {
        if (this->m_opl != NULL && this->m_opr != NULL) {
            return this->m_opl->Compare(pAgent, this->m_opr, this->m_operator);
        } else {
            EBTStatus childStatus = BT_INVALID;
            EBTStatus result = this->update_impl((Agent*)pAgent, childStatus);
            return result == BT_SUCCESS;
        }
    }

    BehaviorTask* Condition::createTask() const {
        ConditionTask* pTask = BEHAVIAC_NEW ConditionTask();

        return pTask;
    }

    void Condition::Cleanup() {
    }

    void ConditionTask::copyto(BehaviorTask* target) const {
        super::copyto(target);
    }

    void ConditionTask::save(IIONode* node) const {
        super::save(node);
    }

    void ConditionTask::load(IIONode* node) {
        super::load(node);
    }

    bool ConditionTask::onenter(Agent* pAgent) {
        BEHAVIAC_UNUSED_VAR(pAgent);
        return true;
    }

    void ConditionTask::onexit(Agent* pAgent, EBTStatus s) {
        BEHAVIAC_UNUSED_VAR(pAgent);
        BEHAVIAC_UNUSED_VAR(s);
    }

    EBTStatus ConditionTask::update(Agent* pAgent, EBTStatus childStatus) {
        BEHAVIAC_UNUSED_VAR(pAgent);
        BEHAVIAC_UNUSED_VAR(childStatus);

        // EBTStatus result = BT_FAILURE;

        BEHAVIAC_ASSERT(Condition::DynamicCast(this->GetNode()));
        Condition* pConditionNode = (Condition*)(this->GetNode());

        bool ret = pConditionNode->Evaluate(pAgent);

        return ret ? BT_SUCCESS : BT_FAILURE;
    }
}

File: tools/designer/Plugins/PluginBehaviac/Nodes/Conditions/Condition.cs
/////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// Tencent is pleased to support the open source community by making behaviac available.
//
// Copyright (C) 2015-2017 THL A29 Limited, a Tencent company. All rights reserved.
//
// Licensed under the BSD 3-Clause License (the "License"); you may not use this file except in compliance with
// the License. You may obtain a copy of the License at http://opensource.org/licenses/BSD-3-Clause
//
// Unless required by applicable law or agreed to in writing, software distributed under the License is
// distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and limitations under the License.
/////////////////////////////////////////////////////////////////////////////////////////////////////////////////

using System;
using System.Collections.Generic;
using System.Text;
using System.Drawing;
using Behaviac.Design;
using Behaviac.Design.Nodes;
using Behaviac.Design.Attributes;
using PluginBehaviac.Properties;

namespace PluginBehaviac.Nodes
{
    [NodeDesc("Conditions:Leaf", NodeIcon.Condition)]
    public class Condition : Behaviac.Design.Nodes.Condition
    {
        public Condition()
        : base(Resources.Condition, Resources.ConditionDesc)
        {
        }

        public override string ExportClass
        {
            get
            {
                return "Condition";
            }
        }

        public override bool HasPrefixLabel
        {
            get
            {
                return false;
            }
        }

        private RightValueDef _opl;
        [DesignerRightValueEnum("OperandLeft", "OperandLeftDesc", "Condition", DesignerProperty.DisplayMode.Parameter, 0, DesignerProperty.DesignerFlags.NoFlags, DesignerPropertyEnum.AllowStyles.AttributesMethod, MethodType.Getter, "", "Opr")]
        public RightValueDef Opl
        {
            get
            {
                return _opl;
            }
            set
            {
                this._opl = value;
            }
        }

        private OperatorType _operator = OperatorType.Equal;
        [DesignerEnum("Operator", "OperatorDesc", "Condition", DesignerProperty.DisplayMode.Parameter, 1, DesignerProperty.DesignerFlags.NoFlags, "ConditionOperaptor")]
        public OperatorType Operator
        {
            get
            {
                return _operator;
            }
            set
            {
                _operator = value;
            }
        }

        private RightValueDef _opr;
        [DesignerRightValueEnum("OperandRight", "OperandRightDesc", "Condition", DesignerProperty.DisplayMode.Parameter, 2, DesignerProperty.DesignerFlags.NoFlags, DesignerPropertyEnum.AllowStyles.ConstAttributesMethod, MethodType.Getter, "Opl", "")]
        public RightValueDef Opr
        {
            get
            {
                return _opr;
            }
            set
            {
                this._opr = value;
            }
        }

        public override Behaviac.Design.ObjectUI.ObjectUIPolicy CreateUIPolicy()
        {
            return new Behaviac.Design.ObjectUI.ConditionUIPolicy();
        }

        public override bool ResetMembers(MetaOperations metaOperation, AgentType agentType, BaseType baseType, MethodDef method, PropertyDef property)
        {
            bool bReset = false;

            if (this.Opl != null)
            {
                bReset |= this.Opl.ResetMembers(metaOperation, agentType, baseType, method, property);
            }

            if (this.Opr != null)
            {
                bReset |= this.Opr.ResetMembers(metaOperation, agentType, baseType, method, property);
            }

            bReset |= base.ResetMembers(metaOperation, agentType, baseType, method, property);

            return bReset;
        }

        public override string Description
        {
            get
            {
                //not ideal, for the left node list
                if (_opl == null && _opr == null)
                {
                    return base.Description;
                }

                string str = base.Description;

                if (_opl != null)
                {
                    str += "\n" + _opl.GetExportValue();
                }

                str += "\n" + _operator.ToString();

                if (_opr != null)
                {
                    str += "\n" + _opr.GetExportValue();
                }

                return str;
            }
        }

        public override object[] GetExcludedEnums(DesignerEnum enumAttr)
        {
            //List<object> excludedOperatorsResult = new List<object>() { OperatorType.Assignment, OperatorType.In };
            List<object> excludedOperatorsResult = new List<object>()
            {
                OperatorType.Assignment
            };

            if (enumAttr != null && enumAttr.ExcludeTag == "ConditionOperaptor")
            {
                if (this.Opl != null)
                {
                    if (this.Opl.ValueType != typeof(bool))
                    {
                        //and and or are only valid for bool, so to exclude and and or when the type is not bool
                        object[] excludedOperators = new object[] { OperatorType.And, OperatorType.Or };
                        excludedOperatorsResult.AddRange(excludedOperators);
                    }
                    else if (this.Opl.ValueType == typeof(bool))
                    {
                        object[] excludedOperators = new object[] { OperatorType.Greater, OperatorType.GreaterEqual, OperatorType.Less, OperatorType.LessEqual };
                        excludedOperatorsResult.AddRange(excludedOperators);
                    }
                }
            }

            return excludedOperatorsResult.ToArray();
        }

        protected override void CloneProperties(Node newnode)
        {
            base.CloneProperties(newnode);

            Condition prec = (Condition)newnode;

            prec._operator = _operator;

            if (_opl != null)
            {
                prec._opl = (RightValueDef)_opl.Clone();
            }

            if (_opr != null)
            {
                prec._opr = (RightValueDef)_opr.Clone();
            }
        }

        public override void CheckForErrors(BehaviorNode rootBehavior, List<ErrorCheck> result)
        {
            if (this.Opl == null || this.Opr == null || this.Opl.ToString() == "" || this.Opr.ToString() == "")
            {
                result.Add(new Node.ErrorCheck(this, ErrorCheckLevel.Error, Resources.OperatandIsNotComplete));
            }
            else if (this.Opl.ValueType != typeof(bool) &&
                     (this.Operator == OperatorType.And || this.Operator == OperatorType.Or))
            {
                // And and Or are only valid for bool
                result.Add(new Node.ErrorCheck(this, ErrorCheckLevel.Error, Resources.AndOrOnlyValidForBool));
            }

            base.CheckForErrors(rootBehavior, result);
        }
    }
}
